Readme.md
[![](https://img.shields.io/npm/v/rn-utilities.svg?style=flat)](https://www.npmjs.com/package/rn-utilities)
[![](https://img.shields.io/npm/dt/rn-utilities.svg)](https://www.npmjs.com/package/rn-utilities)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

# rn-utilities


These are the utility methods for react native to ace the development process.

#### Updated Readme coming soon...

CHANGELOG.md
# [0.2.0](https://github.com/ajaykumar97/rn-utilities/compare/v0.1.11...v0.2.0) (2021-09-19)


### Bug Fixes

* fixed release-it config ([d3000db](https://github.com/ajaykumar97/rn-utilities/commit/d3000db46ca609db761cdbc7be2faee390f9bc35))


### Features

* added `release-it` ([02abb5f](https://github.com/ajaykumar97/rn-utilities/commit/02abb5f633af13e689ce5c2380686fa4882c3bdf))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.8](https://github.com/ajaykumar97/rn-utilities/compare/v0.1.7...v0.1.8) (2021-07-10)


### Features

* Added `conventional-changelog` ([f263e01](https://github.com/ajaykumar97/rn-utilities/commit/f263e01f2a2db839aa23f7b4483bdbf712e92ea4))


src/layout.js
import { Dimensions, Platform, StatusBar } from 'react-native'
import { getStatusBarHeight } from 'react-native-iphone-x-helper';

const width = Dimensions.get('window').width;
const height = Dimensions.get('window').height;

let paddingTop = 0;

const majorVersionOS = parseInt(Platform.Version, 10);

let isOldOsVesion = false;

if (Platform.OS === 'ios') {
    console.log('ios version is:', majorVersionOS);

    if (majorVersionOS < 11) {
        isOldOsVesion = true;
        paddingTop = 20;
    }
} else {
    console.log('android version is:', majorVersionOS);

    if (majorVersionOS < 23) {
        isOldOsVesion = true;
    } else {
        paddingTop = StatusBar.currentHeight;
    }
}

const isIOS = Platform.OS === 'ios';

const layout = {
    size: {
        width,
        height,
    },
    paddingTop,
    statusBarHeight: getStatusBarHeight(),
    isIOS,
    isOldOsVesion
};

export { layout };

src/lib/helpers.js
// util function to convert the input to string type
const convertToString = (input) => {
    if (input) {
        if (typeof input === "string") {
            return input;
        }

        return String(input);
    }

    return '';
}

// convert string to words
const toWords = (input) => {
    input = convertToString(input);

    const regex = /[A-Z\xC0-\xD6\xD8-\xDE]?[a-z\xDF-\xF6\xF8-\xFF]+|[A-Z\xC0-\xD6\xD8-\xDE]+(?![a-z\xDF-\xF6\xF8-\xFF])|\d+/g;

    return input.match(regex);
}

// convert the input array to camel case
const toCamelCase = (inputArray) => {
    let result = '';
    for (let i = 0, len = inputArray.length; i < len; i++) {
        let currentStr = inputArray[i];
        let tempStr = currentStr.toLowerCase();

        if (i != 0) {
            // convert first letter to upper case (the word is in lowercase) 
            tempStr = tempStr.substr(0, 1).toUpperCase() + tempStr.substr(1);
        }

        result += tempStr;
    }

    return result;
}

export {
    convertToString,
    toWords,
    toCamelCase
};


src/helperFunctions.js
const intoTitleCase = (string) => {
    let sentence = string.toLowerCase().split(' ');
    for (let i = 0; i < sentence.length; i++) {
        sentence[i] = sentence[i][0].toUpperCase() + sentence[i].slice(1);
    }
    sentence = sentence.join(' ');
    return sentence;
};

const keyExtractor = (item) => String(item.id);

const getAPIError = (error) => {
    if (
        error.response
        && error.response.data
        && error.response.data.msg
    ) {
        // The request was made and the server responded with a status code
        // that falls out of the range of 2xx

        /* console.log(error.response.data);
        console.log(error.response.status);
        console.log(error.response.headers); */

        return error.response.data.msg;
    }

    return error;
};

const truncateString = (string, numOfChars = 50) => {
    if (string.length > numOfChars) {
        return `${string.slice(0, numOfChars > 3 ? numOfChars - 3 : numOfChars)}...`;
    }
    return str;
};

const getCityRegionCountryFromGoogleApiResults = (addressResults) => {
    let city = {};
    let region = {};
    let country = {};

    for (let j = 0; j < addressResults.length; j++) {
        for (let i = 0; i < addressResults[j].address_components.length; i++) {
            if (addressResults[j].address_components[i].types[0] === 'locality') {
                //this is the object you are looking for City
                city = addressResults[j].address_components[i];
            }
            if (addressResults[j].address_components[i].types[0] === 'administrative_area_level_1') {
                //this is the object you are looking for State
                region = addressResults[j].address_components[i];
            }
            if (addressResults[j].address_components[i].types[0] === 'country') {
                //this is the object you are looking for country
                country = addressResults[j].address_components[i];
            }
        }
    }

    console.log(`${city.long_name} || ${region.long_name} || ${country.short_name}`)

    return {
        city: city.long_name,
        region: region.long_name,
        country: country.short_name
    };
};

//to convert Bytes into KB, MB GB etc
const formatBytes = (bytes, decimals = 2) => {
    if (bytes === 0) return '0 Bytes';

    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
};

//output like: "1,000,000,000"
const formatNumber = (number) => String(number).replace(/(.)(?=(\d{3})+$)/g, '$1,');

//to find intersection between two arrays
const intersectionBy = (a, b, fn) => {
    const s = new Set(b.map(fn));
    return a.filter(x => s.has(fn(x)));
};

const generateRandomNumber = (numberLength = 4) => {
    let multiplier = '9';
    let adder = '1';

    for (let i = 0; i < numberLength - 1; i++) {
        multiplier += '0';
        adder += '0'
    }

    return Math.floor((Math.random() * parseInt(multiplier)) + parseInt(adder));
}

export {
    intoTitleCase,
    keyExtractor,
    getAPIError,
    truncateString,
    getCityRegionCountryFromGoogleApiResults,
    formatBytes,
    formatNumber,
    intersectionBy,
    generateRandomNumber
}

src/functions.js
import { toWords, toCamelCase } from './lib/helpers';

/* 
    To convert normal string into titleCase string, e.g.

    intoTitleCase('tutorix is one of best e-platforms') //Tutorix Is One Of Best E-platforms
*/
const intoTitleCase = (string) => {
    let sentence = string.toLowerCase().split(' ');

    for (let i = 0; i < sentence.length; i++) {
        sentence[i] = sentence[i][0].toUpperCase() + sentence[i].slice(1);
    }
    sentence = sentence.join(' ');
    return sentence;
};

/* 
    To convert normal string into camelCase string, e.g.

    toCamelCaseString('tutorix is one of best e-platforms') //tutorixIsOneOfBestEPlatforms
*/
const toCamelCaseString = (input) => {
    let words = toWords(input);

    return toCamelCase(words);
}

const keyExtractor = (item) => String(item.id);

const getAPIError = (error) => {
    if (
        error.response
        && error.response.data
        && error.response.data.msg
    ) {
        // The request was made and the server responded with a status code
        // that falls out of the range of 2xx

        /* console.log(error.response.data);
        console.log(error.response.status);
        console.log(error.response.headers); */

        return error.response.data.msg;
    }

    return error;
};

const truncateString = (string, numOfChars = 50) => {
    if (string.length > numOfChars) {
        return `${string.slice(0, numOfChars > 3 ? numOfChars - 3 : numOfChars)}...`;
    }
    return str;
};

const getCityRegionCountryFromGoogleApiResults = (addressResults) => {
    let city = {};
    let region = {};
    let country = {};

    for (let j = 0; j < addressResults.length; j++) {
        for (let i = 0; i < addressResults[j].address_components.length; i++) {
            if (addressResults[j].address_components[i].types[0] === 'locality') {
                //this is the object you are looking for City
                city = addressResults[j].address_components[i];
            }
            if (addressResults[j].address_components[i].types[0] === 'administrative_area_level_1') {
                //this is the object you are looking for State
                region = addressResults[j].address_components[i];
            }
            if (addressResults[j].address_components[i].types[0] === 'country') {
                //this is the object you are looking for country
                country = addressResults[j].address_components[i];
            }
        }
    }

    console.log(`${city.long_name} || ${region.long_name} || ${country.short_name}`)

    return {
        city: city.long_name,
        region: region.long_name,
        country: country.short_name
    };
};

//to convert Bytes into KB, MB GB etc
const formatBytes = (bytes, decimals = 2) => {
    if (bytes === 0) return '0 Bytes';

    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
};

/* 
    To format number into comma separated number according to Tens position

    e.g.:
        
        formatNumber(1000000000) //"1,000,000,000"
*/

const formatNumber = (number) => String(number).replace(/(.)(?=(\d{3})+$)/g, '$1,');

/* 
    To find intersection between two arrays 
    e.g.:

    intersectionBy([2.1, 1.2], [2.3, 3.4], Math.floor); // [2.1]
*/
const intersectionBy = (a, b, fn) => {
    const s = new Set(b.map(fn));
    return a.filter(x => s.has(fn(x)));
};

const generateRandomNumber = (numberLength = 4) => {
    let multiplier = '9';
    let adder = '1';

    for (let i = 0; i < numberLength - 1; i++) {
        multiplier += '0';
        adder += '0'
    }

    return Math.floor((Math.random() * parseInt(multiplier)) + parseInt(adder));
}

//to convert RGB color into hex code
const RGBToHex = (r, g, b) => ((r << 16) + (g << 8) + b).toString(16).padStart(6, '0');

/*

    To convert date object into colon format, e.g.
    
    getColonTimeFromDate(new Date()); // "08:38:00"

*/
const getColonTimeFromDate = date => date.toTimeString().slice(0, 8);


/* 
    To check if it is an absolute URL or not, e.g.
    
    isAbsoluteURL('https://google.com'); // true
    isAbsoluteURL('ftp://www.myserver.net'); // true
    isAbsoluteURL('/foo/bar'); // false

*/
const isAbsoluteURL = str => /^[a-z][a-z0-9+.-]*:/.test(str);

/* 

    To compare two objects, e.g.

*/
const isDeepEqual = (obj1, obj2, testPrototypes = false) => {
    if (obj1 === obj2) {
        return true
    }

    if (typeof obj1 === "function" && typeof obj2 === "function") {
        return obj1.toString() === obj2.toString()
    }

    if (obj1 instanceof Date && obj2 instanceof Date) {
        return obj1.getTime() === obj2.getTime()
    }

    if (
        Object.prototype.toString.call(obj1) !==
        Object.prototype.toString.call(obj2) ||
        typeof obj1 !== "object"
    ) {
        return false
    }

    const prototypesAreEqual = testPrototypes
        ? isDeepEqual(
            Object.getPrototypeOf(obj1),
            Object.getPrototypeOf(obj2),
            true
        )
        : true

    const obj1Props = Object.getOwnPropertyNames(obj1)
    const obj2Props = Object.getOwnPropertyNames(obj2)

    return (
        obj1Props.length === obj2Props.length &&
        prototypesAreEqual &&
        obj1Props.every(prop => isDeepEqual(obj1[prop], obj2[prop]))
    )
}

export {
    intoTitleCase,
    toCamelCaseString,
    keyExtractor,
    getAPIError,
    truncateString,
    getCityRegionCountryFromGoogleApiResults,
    formatBytes,
    formatNumber,
    intersectionBy,
    generateRandomNumber,
    RGBToHex,
    getColonTimeFromDate,
    isAbsoluteURL,
    isDeepEqual
}
